### DISCUSSION TITLE
Q: How to access "index" via @JStacheLambda or ...  Q: What is the best way to access the "previous" element?

### DISCUSSION THREAD

**Discussion by rbygrave** [jstachio/jstachio#410] [Q&A]
I'm wondering the best way to access the **_previous element_** when iterating a collection.

For example, I don't think it is supported to access the current iterator `index` value via `@JStacheLambda`? This does not seem to be supported currently, did I miss anything here?
```java
    @JStacheLambda
    public Something previous(int index) {
   ...
```

What we can do to achieve this is to take the current element and find the previous element inside the function like:
```java
    @JStacheLambda
    public Something previous(MyElement current) {
     // find the previous MyElement using the current element myself by iterating the collection
   ...
```

Is there a better way to obtain the "previous" element?

Thanks, Rob.

**Comment by agentgt** [comment]
I’m on mobile but I’m absolutely positive the lambda will have access to the index. 

This is shown in the lambda doc with isEven: https://jstach.io/doc/jstachio/current/apidocs/io.jstach.jstache/io/jstach/jstache/JStacheLambda.html

However that does not get you the last or next without access to the list.

Tomorrow when on a computer I will look at this more.

EDIT for ease of reference here is the `isEven` example:

```hbs
 {{#someList}}
 {{.}} is {{#-index.isEven}}{{#.}}even{{/.}}{{^.}}odd{{/.}}{{/-index.isEven}}
 {{/someList}}
```

`-index` is a virtual key: https://jstach.io/doc/jstachio/current/apidocs/#mustache_virtual_keys_iterable


JStachio supports dotted lambda access which makes `isEven` kind of look like a method on integer (akin to Ruby or some other hardcore oop language) even though it is not.

In fact its expanding notation would be:

```hbs
{{#-index}}{{#isEven}}{{#.}}even{{/.}}{{^.}}odd{{/.}}{{/isEven}}{{/-index}}
```

Because of the way context lookup works `isEven` only needs to be defined higher up the object tree (context tree and lets ignore collisions for now). A typical best practice is to put helper lambda methods as an interface and mixin them into your model:

```java
public interface IterationSupport {
 @JStacheLambda
 default boolean isEven(int index) {
     return index % 2 == 0;
 }
}
```

then have the model implement such interface.

**Comment by agentgt** [comment]
@rbygrave accessing the previous element of a collection is difficult even in more feature rich templating languages like Handlebars or Freemarker.

I am curious of your use case as that has not been a typical requirement unlike knowing first/last or what index for even/odd/n-modula like calculation

The canonical way to do hard things in Mustache is to add additional data usually via decoration.

That is assume we have a `List<String>` where we want your behavior you would have a helper method that would convert that `List<String>` to `List<Item<String>>` where `Item` might look like:

```java
record Item<T>(T current, @Nullable T previous, @Nullable T next) {
}
```

Our model would look like:

```java
@JStache( template = """
{{#listOfItems}}
{{current}}
{{#previous}}
prev = {{.}}
{{/previous}}
{{/listOfItems}}
""")
record Page(List<String> items) {

  List<Item<String>> listOfItems() {
    // use some standard static util function
    return Models.decorate(items);
  }
}
```

Now here is the bad news. There is no way syntactically (and semantically) in Mustache to have a lambda access the entire list at this time (I'll leave that as an exercise to figure out why by the ambiguity of sections). [However there is a proposal to add something currently being called `filters` that would be able to give you access](https://github.com/mustache/spec/discussions/153). I'll ping @jgonggrijp as some more evidence on the potential niceties of filters.

So we are stuck with decorating the model for now. In JStachio this is far less painful than others because the model code is married to the template as well as compile time safety. Because you have to always add a model class IMO it is not too painful to add helpers.

Regardless it is one of the tradeoffs of using a logicless templating language over basically a templating language that is just Java (JSP, JTE, Rocker, etc).

**Comment by agentgt** [comment]
@rbygrave Were you able to resolve your use case?

**Comment by rbygrave** [comment]
> Were you able to resolve your use case?

Yes.

My use case was where I had model classes that ideally would have had both start [date] and end [date] but instead only had one of those attributes [a sub-optimal model design but its old and that's how it is]. I wanted the UI to "drill down" for a given record, which desired both the start and end [date] and where one of those attributes was on the previous record.


> hard things in Mustache is to add additional data usually via decoration.

I'm feeling this is still the correct approach. For my case its that I was developing away and not needing that decoration etc until later on and then going ... ah, it was a decoration that really needed to turn a list of things into a list of pairs of things. So I was looking for a "short cut" for obtaining the previous record instead of adding the decoration.


So yes I'm happy. In terms of thinking about "Logic-less templating", for myself its been a while and I'm rusty but I'm still in that camp where we really want the template to not have logic and make/prepare our "view model" for the template. I'm thinking this question is my getting a bit loose in my thinking to some extent :)


As some feedback, banged up a bunch of Admin UI for an app and really pretty happy with the compile time templating approach. It's pretty cool to compile and know that at least the templating is going to work as expected so time lost in server restarts is made up by that compile time check. Overall feeling really good. Using htmx and the rendering speed browser wise seems impressive and its all stateless so restart server, hit the ui button to get the updated partial page response - not a bad dev experience.

**Comment by agentgt** [comment]
>  It's pretty cool to compile and know that at least the templating is going to work as expected so time lost in server restarts is made up by that compile time check.

There are tricks to this on modern computers and fast application stacks. You just [restart on template change by using some watching daemon](https://github.com/jstachio/jstachio/issues/187#issuecomment-1727725002).

The issue with the above is that there is not a standard way to do this but I have come to terms that restart is the only reliable option because very often you are changing the model (regular Java code) and templates at the same time.

I think a Maven or Gradle plugin could do this watching and restarting but then its at odds with other framework solutions like Jooby's hot reload and Spring Boots devtools.
